tooling/test/sugar.go

```diff
@@ -208,6 +216,67 @@ func (e ExpectBuilder) BodyWithHint(hint string, body interface{}) ExpectBuilder
 	return e
 }
 
+func (e ExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	checks := validateResponse(t, e, res)
+	for _, c := range checks {
+		t.Run(c.testName, func(t *testing.T) {
+			if !c.checkOutput.Success {
+				localReport(t, c.checkOutput.Reason)
+			}
+		})
+	}
+}
+
+type AllOfExpectBuilder struct {
+	Expect_ []ExpectValidator `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AllOfExpectBuilder)(nil)
+
+func AllOf(expect ...ExpectValidator) AllOfExpectBuilder {
+	return AllOfExpectBuilder{Expect_: expect}
+}
+
+func (e AllOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	for _, expect := range e.Expect_ {
+		expect.Validate(t, res, localReport)
+	}
+}
+
+type AnyOfExpectBuilder struct {
+	Expect_ []ExpectBuilder `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AnyOfExpectBuilder)(nil)
+
+func AnyOf(expect ...ExpectBuilder) AnyOfExpectBuilder {
+	return AnyOfExpectBuilder{Expect_: expect}
+}
+
+func (e AnyOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
```

Comment:
We put a lot of emphasis on reporting for this test suite: one of our goals is to have [really clear dashboards](https://singulargarden.github.io/gateway-conformance/tests/testtrustlesscarpathing--get-default-car-response-for-non-existing-file--accept-header-/) that maintainers (and potential users) can check to compare implementations and expectation.

In the case of `AnyOf`, we should make sure that reports are identical between gateways and that the supported features are explicit (range requests support vs naive responses, etc). 

Naively, that function would look something like:

```go
originalFailure := localReport.inCaseOfFailure // assume t.Fail but that might change with recursion
hasSucceeded := false

t.Run(fmt.Sprintf("Check %d", i), {
   if (hasSucceeded || !isLastCheck) {
      localReport.inCaseOfFailure = t.Skip
   } else {
      localReport.inCaseOfFailure = originalFailure
   }

   hasSucceeded = hasSucceeded || expect.Validate(t, res, localReport)    
})

localReport.inCaseOfFailure = originalFailure
return hasSucceeded
```

Reply:
Ah ok, so we want the same number of testcases when possible. Can do.